libyul/optimiser/ReasoningBasedSimplifier.cpp

```diff
@@ -114,224 +153,199 @@ ReasoningBasedSimplifier::ReasoningBasedSimplifier(
+	{
+		// TODO this disallows literals as function arguments.
+		if (!holds_alternative<Identifier>(arg))
+			return;
+		Identifier const& v = get<Identifier>(arg);
+		if (!m_ssaVariables.count(v.name) || !m_variables.count(v.name))
+			return;
+		arguments.push_back(m_variables.at(v.name));
+	}
+
 	switch (_instruction)
 	{
 	case evmasm::Instruction::ADD:
-		return wrap(arguments.at(0) + arguments.at(1));
-	case evmasm::Instruction::MUL:
-		return wrap(arguments.at(0) * arguments.at(1));
+		if (makesInfeasible(arguments.at(0) + arguments.at(1) >= smtutil::Expression(bigint(1) << 256)))
+			m_solver->addAssertion(variable == arguments.at(0) + arguments.at(1));
+		break;
+
 	case evmasm::Instruction::SUB:
-		return wrap(arguments.at(0) - arguments.at(1));
-	case evmasm::Instruction::DIV:
-		return smtutil::Expression::ite(
-			arguments.at(1) == constantValue(0),
-			constantValue(0),
-			arguments.at(0) / arguments.at(1)
-		);
-	case evmasm::Instruction::SDIV:
-		return smtutil::Expression::ite(
-			arguments.at(1) == constantValue(0),
-			constantValue(0),
-			// No `wrap()` needed here, because -2**255 / -1 results
-			// in 2**255 which is "converted" to its two's complement
-			// representation 2**255 in `signedToUnsigned`
-			signedToUnsigned(smtutil::signedDivisionEVM(
-				unsignedToSigned(arguments.at(0)),
-				unsignedToSigned(arguments.at(1))
-			))
-		);
-	case evmasm::Instruction::MOD:
-		return smtutil::Expression::ite(
-			arguments.at(1) == constantValue(0),
-			constantValue(0),
-			arguments.at(0) % arguments.at(1)
-		);
-	case evmasm::Instruction::SMOD:
-		return smtutil::Expression::ite(
-			arguments.at(1) == constantValue(0),
-			constantValue(0),
-			signedToUnsigned(signedModuloEVM(
-				unsignedToSigned(arguments.at(0)),
-				unsignedToSigned(arguments.at(1))
-			))
-		);
+		if (makesInfeasible(arguments.at(0) - arguments.at(1) < 0))
+			m_solver->addAssertion(variable == arguments.at(0) - arguments.at(1));
+		break;
+	//case evmasm::Instruction::MUL:
+		// TODO encode constants?
+	//case evmasm::Instruction::DIV:
 	case evmasm::Instruction::LT:
-		return booleanValue(arguments.at(0) < arguments.at(1));
-	case evmasm::Instruction::SLT:
-		return booleanValue(unsignedToSigned(arguments.at(0)) < unsignedToSigned(arguments.at(1)));
+		m_solver->addAssertion(variable <= 1);
+		m_nonEncodedValues.insert({_varName, arguments.at(0) < arguments.at(1)});
+		break;
 	case evmasm::Instruction::GT:
-		return booleanValue(arguments.at(0) > arguments.at(1));
-	case evmasm::Instruction::SGT:
-		return booleanValue(unsignedToSigned(arguments.at(0)) > unsignedToSigned(arguments.at(1)));
+		m_solver->addAssertion(variable <= 1);
+		m_nonEncodedValues.insert({_varName, arguments.at(0) > arguments.at(1)});
+		break;
 	case evmasm::Instruction::EQ:
-		return booleanValue(arguments.at(0) == arguments.at(1));
+		m_solver->addAssertion(variable <= 1);
+		m_nonEncodedValues.insert({_varName, arguments.at(0) == arguments.at(1)});
+		break;
```

Comment:
```
if (makesInfeasible(arguments.at(0) < arguments.at(1)) && makesInfeasible(arguments.at(0) > arguments.at(1))
    m_solver.add_assertion(variable = 1)
else if (makesInfeasible(arguments.at(0) >= arguments.at(1)) || makesInfeasible(arguments.at(0) <= arguments.at(1))
    m_solver.add_assertion(variable == 0)
else
    m_solver.add_assertion(variable <= 1)
```

This should be able to handle the test case, that is currently failing.

```
let x := 7
let y := 8
if eq(add(x, y), 15) { }
```

Reply:
Ah right!

Reply:
Ah, but we somehow also need to store the relation if it cannot be decided right away, because it might get clearer later. Yeah, this should really be extended to include boolean variables in the solver engine.